### PR TITLE
Refactor expression miner interface

### PR DESCRIPTION
**Pull request by ajreynol**
This is work towards a major refactor of our algorithms for expression mining (rewrite rule synthesis and verification, query generation).  The plan is to make these available via a `find-synth` command, see the changes in the regressions for examples.  Currently, these algorithms are available via options that "hijack" calls to `check-sat`/`check-synth`, which is highly non-intuitive and requires modifications to the core of the SyGuS solver.

This changes the interface to expression miner in preparation for this change.

Note this currently removes our support for expression mining via options (which is marked as expert options only).  Support for these will be available again when support for `find-synth` is completed.  Several regressions are disabled in the meantime.

### CODE REVIEW

File: src/theory/quantifiers/candidate_rewrite_database.h [hunk 1]
```diff
@@ -125,6 +121,8 @@ class CandidateRewriteDatabase : public ExprMiner
   bool d_filterPairs;
   /** whether we are using sygus */
   bool d_using_sygus;
+  /** Whether we are check rewrite rules for all subterms added to this class */
```
Comment (abdoo8080) [inline src/theory/quantifiers/candidate_rewrite_database.h:124]:
```suggestion
  /** Whether we check rewrite rules for all subterms added to this class */
```

File: src/theory/quantifiers/candidate_rewrite_database.cpp [hunk 2]
```diff
@@ -32,16 +32,23 @@ namespace cvc5::internal {
 namespace theory {
 namespace quantifiers {
 
+/** Attribute true for rewrites where verification was run  */
+struct RewriteVerifiedAttributeId
+{
+};
+typedef expr::Attribute<RewriteVerifiedAttributeId, bool>
+    RewriteVerifiedAttribute;
```
Comment (abdoo8080) [inline src/theory/quantifiers/candidate_rewrite_database.cpp:40]: Is this needed? How about using `std::unordered_set<Node> d_verified` instead?

File: src/theory/quantifiers/expr_miner.h [hunk 3]
```diff
@@ -57,12 +59,10 @@ class ExprMiner : protected EnvObj
                           SygusSampler* ss = nullptr);
   /** add term
    *
-   * This registers term n with this expression miner. The output stream out
-   * is provided as an argument for the purposes of outputting the result of
-   * the expression mining done by this class. For example, candidate-rewrite
-   * output is printed on out by the candidate rewrite generator miner.
+   * This registers term n with this expression miner, and adds an expressions
+   * found (e.g. rewrites, queries) to found.
```
Comment (abdoo8080) [inline src/theory/quantifiers/expr_miner.h:63]:
```suggestion
   * This registers term n with this expression miner, and adds expressions found
   * (e.g. rewrites, queries) to found.
```

File: src/theory/quantifiers/sygus/rcons_type_info.cpp [hunk 4]
```diff
@@ -64,8 +64,8 @@ Node RConsTypeInfo::nextEnum()
 
 Node RConsTypeInfo::addTerm(Node n)
 {
-  std::stringstream out;
-  return d_crd->addTerm(n, false, out);
+  std::vector<Node> rewrites;
+  return d_crd->addOrGetTerm(n, rewrites);
```
Comment (abdoo8080) [inline src/theory/quantifiers/sygus/rcons_type_info.cpp:68]:
```suggestion
  return d_crd->addOrGetTerm(n, {});
```

**Review by abdoo8080 — CHANGES_REQUESTED** [review]

File: src/theory/quantifiers/sygus/rcons_type_info.cpp (hunk 4 above)
Comment (ajreynol) [reply src/theory/quantifiers/sygus/rcons_type_info.cpp:68]: This is a pass by non-const reference, so this change isn't feasible.

File: src/theory/quantifiers/candidate_rewrite_database.cpp (hunk 2 above)
Comment (ajreynol) [reply src/theory/quantifiers/candidate_rewrite_database.cpp:40]: Yes, I agree.

**Review by abdoo8080 — APPROVED** [review]

**Pull request merged**